File: Graph/programmers_네트워크.cpp
//https://programmers.co.kr/learn/courses/30/lessons/43162
#include <string>
#include <vector>
#include <iostream>
using namespace std;
bool visited[200];

void dfs(int start, int n, vector<vector<int>> &computers){
visited[start]=true;
    for(int i=0; i<computers[start].size();i++){
        if(i!=start&&!visited[i]&&computers[start][i]){

            dfs(i,n,computers);
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer=0;
    for(int i=0; i<computers.size();i++){
        for(int j=0; j<computers[i].size();j++){
           // cout << computers[i][j]<<' ';
            if(computers[i][j]){
                cout <<i <<' '<<j<<'\n';
            }
        }
        cout <<'\n';
    }
 
    for(int i=0; i<computers.size();i++){
        if(!visited[i]){
      //      cout <<i<<'\n';
           
        dfs(i,n,computers);
            answer++;
        }
    }
    dfs(0,n,computers);
    return answer;
}

File: Graph/programmers_네트워크.kt
//https://programmers.co.kr/learn/courses/30/lessons/43162
class Solution {
    
    fun dfs(visited: BooleanArray, computers : Array<IntArray>,start : Int){
        visited[start]=true
        for(i in computers[start].indices){
            if(computers[start][i]==1 && !visited[i]){
                dfs(visited,computers,i)
            }
        }
    }
    
    fun solution(n: Int, computers: Array<IntArray>): Int {
        var answer = 0
        val visited = BooleanArray(n){false}
        
        for(i in computers.indices){
            if(!visited[i]){
                dfs(visited,computers,i)
                answer++
            }
        }
        return answer
    }
}
